File: CPP/Lesson_094_find_critical_and_pseudo_critical_edges_in_mst.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 094 -- Find Critical and Pseudo Critical Edges in MST
// Category   : Advanced Graphs
// Difficulty : Hard
// Study Plan : Day 47
// =============================================================
//
// QUESTION:
//   Given n nodes and weighted edges, find indices of critical and pseudo-critical edges in any MST.
// =============================================================

#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
struct DSU { vector<int> p,r; int cnt;
    DSU(int n):p(n),r(n,0),cnt(n){ iota(p.begin(),p.end(),0); }
    int f(int x){ while(p[x]!=x){ p[x]=p[p[x]]; x=p[x]; } return x; }
    bool u(int a,int b){ a=f(a); b=f(b); if(a==b) return false;
        if(r[a]<r[b]) swap(a,b); p[b]=a; if(r[a]==r[b]) r[a]++; cnt--; return true; }
};
class Solution { public:
    int n; vector<vector<int>> E;
    int kruskal(int skip, int force){
        DSU d(n); int w=0;
        if (force>=0){ if(!d.u(E[force][0],E[force][1])) return INT_MAX; w+=E[force][2]; }
        for (int i=0;i<(int)E.size();i++){ if(i==skip) continue; if(d.u(E[i][0],E[i][1])) w+=E[i][2]; }
        return d.cnt==1 ? w : INT_MAX;
    }
    vector<vector<int>> findCriticalAndPseudoCriticalEdges(int N, vector<vector<int>>& edges){
        n=N; E=edges; for (int i=0;i<(int)E.size();i++) E[i].push_back(i);
        sort(E.begin(),E.end(),[](auto&a,auto&b){return a[2]<b[2];});
        int base=kruskal(-1,-1); vector<int> crit, pseu;
        for (int i=0;i<(int)E.size();i++){
            if (kruskal(i,-1) > base) crit.push_back(E[i][3]);
            else if (kruskal(-1,i) == base) pseu.push_back(E[i][3]);
        }
        return {crit, pseu};
    }
};
int main(){ vector<vector<int>> e={{0,1,1},{1,2,1},{2,3,2},{0,3,2},{0,4,3},{3,4,3},{1,4,6}};
    auto r=Solution().findCriticalAndPseudoCriticalEdges(5,e);
    for (auto& v: r){ cout<<"["; for(int x: v)cout<<x<<","; cout<<"] "; } cout<<endl; }

File: JavaScript/Lesson_094_find_critical_and_pseudo_critical_edges_in_mst.js
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 094 -- Find Critical and Pseudo Critical Edges in MST
// Category   : Advanced Graphs
// Difficulty : Hard
// Study Plan : Day 47
// =============================================================
//
// QUESTION:
//   Given n nodes and weighted edges, find indices of critical and pseudo-critical edges in any MST.
// =============================================================

class DSU{ constructor(n){this.p=Array(n).fill(0).map((_,i)=>i); this.r=Array(n).fill(0); this.cnt=n;}
  f(x){ while(this.p[x]!==x){this.p[x]=this.p[this.p[x]]; x=this.p[x];} return x; }
  u(a,b){ a=this.f(a); b=this.f(b); if(a===b) return false;
    if(this.r[a]<this.r[b]){[a,b]=[b,a];} this.p[b]=a;
    if(this.r[a]===this.r[b]) this.r[a]++; this.cnt--; return true;}
}
var findCriticalAndPseudoCriticalEdges = function(n, edges){
  edges.forEach((e,i)=>e.push(i));
  edges.sort((a,b)=>a[2]-b[2]);
  const k=(skip=-1,force=-1)=>{
    const d=new DSU(n); let w=0;
    if (force>=0){ const [u,v,wt]=edges[force]; if(!d.u(u,v)) return Infinity; w+=wt; }
    for (let i=0;i<edges.length;i++){ if(i===skip) continue; const [u,v,wt]=edges[i]; if(d.u(u,v)) w+=wt; }
    return d.cnt===1 ? w : Infinity;
  };
  const base=k(); const crit=[], pseu=[];
  for (let i=0;i<edges.length;i++){
    if (k(i,-1) > base) crit.push(edges[i][3]);
    else if (k(-1,i) === base) pseu.push(edges[i][3]);
  }
  return [crit, pseu];
};
console.log(JSON.stringify(findCriticalAndPseudoCriticalEdges(5, [[0,1,1],[1,2,1],[2,3,2],[0,3,2],[0,4,3],[3,4,3],[1,4,6]])));
